xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+    if (Action.DENY.equals(authConfig.action) == (firstMatch == null)) {
+      decisionType = Action.ALLOW;
+    }
+    log.log(Level.FINER, String.format("RBAC decision: {%s}, policy match: {%s}",
+        decisionType, firstMatch));
+    return AuthDecision.create(decisionType, firstMatch);
+  }
+
+  /**
+   * Implements a top level {@link Matcher} for a single RBAC policy configuration per envoy
+   * protocol:
+   * https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#config-rbac-v3-policy.
+   *
+   * <p>Currently we only support matching some of the request fields. Those unsupported fields are
+   * considered not match until we stop ignoring them.
+   */
+  static final class PolicyMatcher extends Matcher {
+    private final OrMatcher permissions;
+    private final OrMatcher principals;
+    private final String name;
+
+    /** Constructs a matcher for one RBAC policy. */
+    PolicyMatcher(String name, OrMatcher permissions, OrMatcher principals) {
+      this.name = name;
+      this.permissions = permissions;
+      this.principals = principals;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return permissions.matches(args) && principals.matches(args);
+    }
+  }
+
+  static final class AuthenticatedMatcher extends Matcher {
+    private final StringMatcher delegate;
+
+    AuthenticatedMatcher(StringMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      if (delegate == null) {
+        return true;
+      }
+      Collection<String> principalNames = args.getPrincipalNames();
+      log.log(Level.FINER, "Matching principal names: " + principalNames);
+      if (principalNames != null) {
+        for (String name : principalNames) {
+          if (delegate.matches(name)) {
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+  }
+
+  static final class DestinationIpMatcher extends Matcher {
+    private final CidrMatcher delegate;
+
+    DestinationIpMatcher(CidrMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getDestinationIp());
+    }
+  }
+
+  static final class SourceIpMatcher extends Matcher {
+    private final CidrMatcher delegate;
+
+    SourceIpMatcher(CidrMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getSourceIp());
+    }
+  }
+
+  static final class PathMatcher extends Matcher {
+    private final StringMatcher delegate;
+
+    PathMatcher(StringMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getPath());
+    }
+  }
+
+  static final class HeaderMatcher extends Matcher {
+    private final Matchers.HeaderMatcher delegate;
```

Comment:
I am actually OK with such a composition for code reuse of `Matchers.HeaderMatcher`'s matching logic for client side name resolver and server side authorization. It looks like an adaptor for server authorization's use case. But could we rename things to avoid conflicts? Maybe `AuthHeaderMatcherAdaptor` or something better?

Reply:
`AuthHeaderMatcher` might be good. I don't care. `-adaptor` seems controversy because I myself am not sophisticated enough to understand it, so i guess other readers may have the same problem.